### ISSUE TITLE
Error authenticating as GitHub application - 'Expiration time' claim ('exp') is too far in the future

### ISSUE THREAD

**Issue by jonmilsom** [GrahamCampbell/Laravel-GitHub#126]
Hi,

Every-so-often I get this error when trying to create an installation token for a GitHub App.

```
$client = GitHubFacade::connection('private');
$client->apps()->createInstallationToken($github_app_installation_id);
```

GitHub specify a _maximum_ expires value of 10 minutes:
https://docs.github.com/en/free-pro-team@latest/developers/apps/authenticating-with-github-apps#authenticating-as-a-github-app

The underlying code in this packages specifies a 9 minute 59 seconds value:
https://github.com/GrahamCampbell/Laravel-GitHub/blob/10.0/src/Auth/Authenticator/PrivateKeyAuthenticator.php#L76

```
        $expires = $issued->add(
            new DateInterval('PT9M59S')
        );
```

My hypothesis is that (despite syncing with ntp) the internal clock on my server is drifting to >+1 seconds from UTC, resulting in a rejected expires value.

As the use of this token is short lived, would you consider a PR to change this value to 9 minutes and 50 seconds? This would be less demanding on the accuracy between systems, but still expose larger discrepancies due to mis-configuration.

e.g.
```
        $expires = $issued->add(
            new DateInterval('PT9M50S')
        );
```

Thank you for reading this issue. I am of course open to other suggestions!

**Comment by GrahamCampbell** [comment]
Possibly we should just backdate the generated at timestamp by 1 minute, and then make the expires time exactly 10 minutes after that? Do you think that could cause any issues? Does GitHub validate if the generated at time is impossible due to the secret key generation time?

**Comment by jonmilsom** [comment]
I have tested the following:

1. Manually adjusted the "expires" time to be more than 10 minutes (forced failure) i.e. `PT9M59S` => `PT10M59S`

Result: Expected error as above "'Expiration time' claim ('exp') is too far in the future"

2. Manually adjusted the clock on my server to be 5 minutes in the past

Result: This worked without error

I would therefore conclude that the `$issued` timestamp is not validated (or if it is, 5 mins is within tolerance)

(I re-tested with the clock set back an hour, and worryingly? this was also accepted....)

--------------------

It would seem unnecessarily risky to change the `$issued` timestamp (GitHub may choose to validate this more strictly in future)... also not very neat. Reducing the `expires` value so that it is not so close the 10 minute expiry limit, seems to have little negative consequence.

Please let me know if I have mis-understood your response.

**Comment by GrahamCampbell** [comment]
Did you test setting the issued at time older than the datestamp of the private key?

**Comment by jonmilsom** [comment]
Apologies - it seems my previous tests were incorrect (internal clock kept correcting itself silently) - please disregard.

Re-testing:
1. Manually adjusted the "expires" time to be more than 10 minutes (forced failure) i.e. PT9M59S => PT10M59S
Result: 'Expiration time' claim ('exp') is too far in the future"

2. Manually adjusted the clock on my server to be 5 minutes in the past
Result: This worked without error

3. Manually adjust the clock on my server to be 1 hour in the past
Result: 'Expiration time' claim ('exp') must be a numeric value representing the future time at which the assertion expires

4. Manually specify `$issued` as one year ago (before private key generated), `$expires` 5 mins in the future
Result: Works

5. Manually specify `$issued` as one year ago (before private key generated), `$expires` +5 mins from `$issued`
Result: 'Expiration time' claim ('exp') must be a numeric value representing the future time at which the assertion expires

--------------------------
So it seems that the validation is on the `$expires` and this must be (a) in the future but (b) 10 minutes or less from now

There does not seem to be any validation on the `$issued` time vs the creation time of the private key.

**Comment by GrahamCampbell** [comment]
Excellent, so there should be no risk in the approach of setting the issued at time 1 minute in the past, and the expires at 10 minutes after the issued at time?

**Comment by jonmilsom** [comment]
This seems to work from my testing:

Code:
```
        $issued = new DateTimeImmutable('-1 minute');

        $expires = $issued->add(
            new DateInterval('PT10M0S')
        );
```

Log:
`[2021-01-06 16:52:25] local.INFO: GrahamCampbell\GitHub\Auth\Authenticator\PrivateKeyAuthenticator {"issued":"2021-01-06 16:51:25","expires":"2021-01-06 17:01:25"}`

Why would you prefer to adjust the `$issued` rather than `$expires`?

**Comment by GrahamCampbell** [comment]
I have gone with adjusting the issued at time to 30 seconds in the past, and the expires at time to exactly 10 minutes after the issued at time. This is implemented in both versions v9.8.1 and 10.0.1. 🚀

**Comment by jonmilsom** [comment]
Thanks @GrahamCampbell 🙌
